coco autoload .dsk and .cas support

udance4ever opened the pull request:
Color Computer (coco) .cas and .dsk files previously did not autoload like other systems (eg. apple2)

Drivers coco & coco3 as of MAME v0.261 have support for .dsk:
```
[root@BATOCERA]# /usr/bin/mame/mame -listmedia coco
SYSTEM           MEDIA NAME       (brief)    IMAGE FILE EXTENSIONS SUPPORTED
---------------- --------------------------- -------------------------------
coco             floppydisk1      (flop1)    .mfi  .dfi  .hfe  .mfm  .td0  .imd  .d77  .d88  .1dd  .cqm  .cqi  .dsk  .dmk  .sdf  .jvc  .vdk  .os9  

[root@BATOCERA]# /usr/bin/mame/mame -listmedia coco3
SYSTEM           MEDIA NAME       (brief)    IMAGE FILE EXTENSIONS SUPPORTED
---------------- --------------------------- -------------------------------
coco3             floppydisk1      (flop1)    .mfi  .dfi  .hfe  .mfm  .td0  .imd  .d77  .d88  .1dd  .cqm  .cqi  .dsk  .dmk  .sdf  .jvc  .vdk  .os9  
```

### Commit Summary
1. allow ES to accept .dsk extension. Add “Disk” to coco altRomType choice 
2. add "Tandy Radio Shack Color Computer cassettes" to coco softList choice in Advanced Game Options (lr-mame, standalone MAME)
3. implement autoload behavior to use “usage” info in coco MAME software lists
4. implement autoload behavior for .cas and .dsk (.bas in rom basename uses CLOAD/RUN)
5. implement autoRunCmd override for cass and flop

### Default Autoload Behavior (top to bottom)
1. "usage" info field when using coco MAME software lists (if declared)
2. Cassette (.cas) files autoload as binary - verified and tested Zaxxon (Datasoft).cas, Zonx (The Rainbow).cas autoload using `CLOADM:EXEC` (F10 is useful to unthrottle and speed up loading :)
6. Disk (.dsk) files autoload based on basename (eg. ZONX.dsk autoloads as `LOADM “ZONX”:EXEC`)
7. BASIC programs autorun if file basename ends with .bas - examples:
4a. ALPHAII.BAS.dsk autoloads using `RUN “ALPHAII”`
4b. GROVER.bas.cas  autoloads using `CLOAD:RUN`
5. user defined override (section below)

### Example Autoload Behavior Found in MAME Software Lists
1. run menu on disks that are a collection of programs (eg. Rainbow on Disk (eg. [rbwd0288](https://github.com/batocera-linux/batocera.linux/blob/6cfa2ca2f041cc7cd9872b62547e54fb94a3792d/package/batocera/core/batocera-configgen/data/mame/coco_flop_autoload.csv#L2)) autoruns `RUN “MENU”`)
3. roms that require a series of commands to autoload (eg. [Sam Sleuth](https://github.com/batocera-linux/batocera.linux/blob/6cfa2ca2f041cc7cd9872b62547e54fb94a3792d/package/batocera/core/batocera-configgen/data/mame/coco_flop_autoload.csv#L1) autoruns `LOADM "SLUTHIGH":EXEC:LOADM "SLUTHLOW":EXEC &H073F` (as specified by developer)
4. cassettes (or disks) that have BASIC bootloaders (eg. [Grover’s Number Rover (CCW).cas](https://github.com/batocera-linux/batocera.linux/blob/6cfa2ca2f041cc7cd9872b62547e54fb94a3792d/package/batocera/core/batocera-configgen/data/mame/coco_cass_autoload.csv#L1) requires `CLOAD:RUN` prior bootloader loading 2nd binary segment on tape into memory)

NOTE: A software list in “Advanced Game Options” **must** be selected to enable ES to search for your rom (and usage info, if available). This is not the default (unless you have set a software list for Color Computer in “Per System Advanced Configuration”)

### User definable overrides
define autoload overrides in: `system/configs/mame/autoload/coco_{cass,flop}_autoload.csv` - examples:
1. run a single BASIC program (able to symlink`1988_02a.dsk` to [Castle of Death (The Rainbow)](https://github.com/batocera-linux/batocera.linux/blob/6cfa2ca2f041cc7cd9872b62547e54fb94a3792d/package/batocera/core/batocera-configgen/data/mame/coco_flop_autoload.csv#L3) and autorun `RUN “CASTLE”`)

blank lines and lines starting with `#` (comments) are skipped in override file.

please consider contributing your overrides (and declaring new software) in our [MAME software lists](https://github.com/mamedev/mame/tree/master/hash)

### Not Implemented
1. tapes that require loading on both sides - user must attach 2nd .cas in MAME File Manager and continue CLOADM (eg. Sam Sleuth (part {1,2}) (Computerware).cas)) (possible future m3u implementation) (.dsk exists so not a high priority)

Verified carts continue to boot (eg. Personal Finance: `finance2.zip`)

If any other coco users want to pitch in and test their coco libraries, this would be great but I think I’ve covered most cases.

Let me know if you have any questions!

### Complete List of Titles Verified to Autoload
1. Rainbow on Disk (Feb 1988) (softList:coco_flop, rbwd0288.zip)
2. Rainbow on Disk (Feb 1988) (1988_02a.dsk, bas, coco_flop override)
2a. Castle of Death (The Rainbow, Feb 1988) (symlink, Castle of Death (The Rainbow).dsk, bas, coco_flop override)
2b The Controllers (The Rainbow) (symlink, CONTROL.bas.dsk)
3. Rainbow on Tape (Apr 1986) (ROT8604a.dsk, bas, coco_flop override)
3a. Maze of Moycullen (symlink, MOYCULEN.BAS.dsk)
4. CoCo Zone (cocozone.dsk, bas, override) (softList:coco_flop, cocozone.zip)
5. Rescue on Alpha II (The Rainbow) (softList:coco_flop, ralpha2.zip)
6. Rainbow Book of Adventures (softList:coco_flop, rnbadv1.zip)
7. Rainbow Book of Adventures (RNBADV1{a,b}.dsk, bas, coco_flop override)
7a. Deed of the York (symlink, DEEDYORK.BAS.dsk)
7b. Dungeon Adventure (symlink, DUNGEON.BAS.dsk)
7c. Escape from Sparta (symlink, ESSPARTA.BAS.dsk)
7d. Head of the Beast (symlink, HEAD.BAS.dsk)
7e. Lighthouse Adventure (symlink, LIGHTADV.BAS.dsk)
7f. Search for the Ruby Chalice (symlink, RCHALICE.BAS.dsk)
7g. An Unexplored Mansion (symlink, UMANSION.BAS.dsk)
8. Sir Randolf of the Moors (SRANDOLF.dsk, bas)
9. Second Rainbow Book of Adventures (softList:coco_flop, rnbadv2.zip)
10. CoCo Max III (v3.1) (softList:coco_flop, ccmax31.zip)
11. Grover’s Number Rover (softList:coco_flop, grover.zip)
12. Grover’s Number Rover (cas/grover.zip, coco_cass override) { prep for softList: coco_cass }
13. JDOS v1.23 (ccc)
14. Lansford Mansion (softList:coco_flop, lansford.zip)
15. Personal Finance II (softList:coco_cart, finance2.zip)
16. Sam Sleuth (Computerware) (softList:coco_flop, ssleuth.zip)
17. Zaxxon (Datasoft) (cas, bin)
18. Zaxxon (Datasoft) (wav, bin)
19. Zaxxon (Datasoft) (cas/zaxxon.zip) { prep for softLlist: coco_cass }
20. Zaxxon (Datasoft) (softList:coco_flop, zaxxon.zip)
21. Zonx (The Rainbow) (cas, bin) 
22. Zonx (The Rainbow) (cas/zonx.zip, coco_cass override) { prep for softList: coco_cass }
23. Zonx (The Rainbow) (softList:coco_flop, zonx.zip)

dmanlfc commented:
squash your commits

`- Override behavior in /usr/share/batocera/configgen/data/mame/coco_{cass,flop}_autoload.csv`  - if these files have the intention to be user configurable, then it should be in `/userdata/configs/mame/` etc.

udance4ever commented:
> squash your commits

Do you have a good article on how to squash commits?

I tried practicing on a [simpler branch in mamedev](https://github.com/mamedev/mame/pull/12378) and feel it made it less readable by bringing non-related commits into the branch.

> Override behavior in `/usr/share/batocera/configgen/data/mame/coco_{cass,flop}_autoload.csv` - if these files have the intention to be user configurable, then it should be in `/userdata/configs/mame/` etc.

sanity check: does configgen ever read in XML software lists?

I ask because I believe it is more elegant to extract the `usage` statements already in the software lists and avoid duplicate effort:
```
<info name="usage" value="LOADM&quot;ZONX&quot;:EXEC" />
```

dmanlfc commented:
I'm not sure you the software that you use but there are numerous ways to do it. Based on how you're working read the associated article(s).
`configgen` is just python & therefore flexible to leverage various xml source files to extract what you need to improve the user experience before launching etc.

udance4ever commented:
> I'm not sure of the software that you use but there are numerous ways to do it. Based on how you're working read the associated article(s).

did you mean to link some articles?    I’m just using CLI git and [this stackexchange article](https://stackoverflow.com/a/24310701/9983389) makes it sound like this is the right command:

```
git rebase -i master
```

it looked like it was working using `pick` and `squash` until I wanted to push my commits - git complained I needed to pull and next thing my branch is cluttered with commits from other contributors.  I don’t want a rerun of that here (it’s already happened to me twice)

if you have a favorite tool/process you like to use, I’m happy to try it.

> `configgen` is just python & therefore flexible to leverage various xml source files to extract what you need to improve the user experience before launching etc. The XML files obviously need to be present beforehand.

ok based on your response, it doesn’t sound like there’s a global object anywhere with this data.  I got the hang of ElementTree this past week so this is what I’d use.

Don’t merge anything until I give the green light - it’s okay if this gets included in v41 (i’ll update the changelog as needed)

I plan to add a menu option for the coco Cassette Software List so I want to build Emulation Station from source to make sure it works.

dmanlfc commented:
I use sublime merge. but `git rebase -i HEAD~14` will do you last 14 commits. and then squash accordingly. remember to use --force. i.e. `git push origin <coco_boot> --force`. then use `--force` for subsequent updates.

udance4ever commented:
> `git push origin coco_boot --force`

THANK YOU for that - I really struggled with that and this worked perfectly!  so much cleaner and now I have no fear of squashing commits in the future ☺️

I set up my build env using the [build instructions in our wiki](https://wiki.batocera.org/compile_batocera.linux) - had no issues w Docker and `make x86_64-buildclean` fired off for about 1.5 hours and running into an issue compiling `host-batocera-llvm-18.1.2` - here's the output from my build: [make.x86_64-cleanbuild.log.gz](https://github.com/batocera-linux/batocera.linux/files/15344815/make.x86_64-cleanbuild.log.gz)

While my long term goal is to build a Batocera Linux image from source, is there a way short-term to rebuild just `/usr/bin/emulationstation` so I can validate the menu changes?  I tried to go into the shell using `make x86_64-shell` and `make batocera-emulationstation` but this target insists on building the llvm target in the build container (btw, as an integrator in a past life, the use of containers for builds is a really neat innovation and can really appreciate how many cross-platform build problems we used to have back in the day!)

udance4ever commented:
> if these files have the intention to be user configurable, then it should be in `/userdata/configs/mame/` etc.

ok - I’m done incorporating your suggestion and pretty happy with the result!

1. I added & tested 8 coco disk roms I contributed to the [official coco_flop MAME software list](https://github.com/mamedev/mame/pull/12386)
2. optional overrides now definable in  `system/configs/mame/autoload`

updated the changelog & squashed my commits :)

only thing remaining is getting my hands on build of ES that contains the menu changes.  Might you have a build server in the cloud to offer I can get permission to use?

dmanlfc commented:
are these files still necessary - coco_{cass,flop}_autoload.csv ?
if so, you may want to add them to datainit so they're populated in /userdata at install / update.

udance4ever commented:
because the MAME software list can do all the heavylifting, it’s only there for power users who aren’t happy with the defaults.  the files are optional and don’t need to be included on install/update.

dmanlfc commented:
Understood but my only reservation is that you don't provide an example of the csv. So even to an advanced user, it's not clear what should be in the csv file to make modifications. So my advice is to either provide example files, outline what one looks like in the _info.txt file or remove the reference.

udance4ever commented:
thank you very much for the feedback!  I added two files (one simple and one complex example) - I noticed the datainit tree is pretty sparse, did I put them in the right place?

dmanlfc commented:
you need to do it from the mame package, not the `batocera-datainit` package. see `iortcw.mk` or `libreretro-mame2003` or just search `$(TARGET_DIR)/usr/share/batocera/datainit/system/configs` to see multiple package populating the folder accordingly.

udance4ever commented:
thanks for the merge!  tested latest butterfly build (`40-dev-9a61198db7 2024/05/27 01:56`) and everything that **must** work works!

Only 2 minor issues:
1. New coco cassette software list (which now shows up in ES) produces the following error from MAME:
```

evmapy: no process found
2024-05-30 18:08:42,957 ERROR (emulatorlauncher:555):runCommand 
"zonx1" approximately matches the following
supported software items (best match first):
```
I’ll do some troubleshooting to see if ES passes `coco_cass` into `softList` in mameGenerator.  If I had to guess, MAME may need to be enhanced to recognize new software lists.

2. `system/configs/mame/autoload` files don’t show up in a Batocera upgrade - is there an upgrade hook to copy these files into `/userdata/system`?  I notice they exist in the overlay here so they were copied successfully during the build:
```
/overlay/base/usr/share/batocera/datainit/system/configs/mame/autoload
```

udance4ever commented:
> 1. Choosing the new coco cassette software list option in ES produces the following error from MAME (doesn’t find any matches nor does it report any close ones)

just a quick update on this front: one of the MAME contributors was kind to point out 2 new lines of code I needed to add so MAME recognizes the `coco_cass` software list. Change is in the pipe and expect it to be part of MAME build in Batocera v41 (mamedev/mame#12505)
